### DISCUSSION TITLE
Matlab s-function linking - libraries in Visual Studio

### DISCUSSION THREAD

**Discussion by sthelia** [JSBSim-Team/jsbsim#405] [Support requests]
Hi again! 
For those who read my previous posts, my thesis is moving along nicely. I had to change my approach a bit and am now doing a Simulink interface using S-functions. It works nicely on my own Linux computer, and I am now trying to compile it on a Windows computer. (This is inspired by [this JSBSim-Matlab](https://github.com/podhrmic/JSBSim-Matlab) repo, I will post my solution once I have cleaned it up a bit) 

My problem now is on the Windows computer. Since I am using S-functions, I use the `mex` command in matlab. It looked like this for my Linuz computer: 
`mex -v -R2017b CXXFLAGS='$CXXFLAGS -std=c++14' ./matlab/JSBSim_SFunction.cpp ./matlab/JSBSimInterface.cpp -I./src -L./src/.libs -lJSBSim`

I am having trouble with the library files. I have tried to use Visual Studio 2019 and follow the JSBSim [user guide](https://jsbsim-team.github.io/jsbsim-reference-manual/mypages/quickstart-building-using-visualstudio/), but it looks a bit different now than the screenshots in the user guide. I managed to build _something_, but I think it is a debug version because when I run my mex command I get these errors:
``` 
Error using mex
JSBSim.lib(sg_path.cxx.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '2' doesn't match value '0' in
JSBSim_SFunction.obj
JSBSim.lib(sg_path.cxx.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MDd_DynamicDebug' doesn't match value
'MD_DynamicRelease' in JSBSim_SFunction.obj
JSBSim.lib(FGFDMExec.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '2' doesn't match value '0' in
JSBSim_SFunction.obj
JSBSim.lib(FGFDMExec.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MDd_DynamicDebug' doesn't match value
'MD_DynamicRelease' in JSBSim_SFunction.obj
JSBSim.lib(FGPropertyManager.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '2' doesn't match value '0' in
JSBSim_SFunction.obj
JSBSim.lib(FGPropertyManager.cpp.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MDd_DynamicDebug' doesn't match value
'MD_DynamicRelease' in JSBSim_SFunction.obj
JSBSim.lib(FGMatrix33.cpp.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '2' doesn't match value '0' in
JSBSim_SFunction.obj

```
Google told me this has something to do with release vs debug configurations. Currently I'm locating them in the path `jsbsim\out\install\x64-Debug (default)\lib` since that is the only place I found lib-files, but I guess that is part of my problem since it's debug. 

 Does anyone have some advice on how I link to (or create?) the correct library file on Windows? 
Thanks in advance!

**Comment by seanmcleod** [comment]
Yep, it looks like you've built the JSBSim_SFunction code as a release build and JSBSim has been built as a debug build.

So either build the SFunction code as a debug build or build JSBSim as a release build and link against the release version of the .lib.

**Comment by sthelia** [comment]
Right, that is what I'm struggling with. I tried to figure out how to build it as a release, and a "Release"-directory did show up in the JSBSim directory. But I can't locate a .lib there. I've tried to play around a bit in Visual Studio to find how to correctly build a release build, but no luck. 
So my questions are: 
* How do I do a release build in Visual Studio 2019? 
* Where does the lib-file end up? 

Thanks for your reply!

**Comment by seanmcleod** [comment]
Okay, so one potentially confusing issue is that we have 2 ways of building JSBSim using Visual Studio. 

The first option is to make use of the provided Visual Studio solution/project file to build JSBSim. Which is what the majority of Visual Studio users are most familiar with. However using this option the project file doesn't generate a .lib file, it simply generates an .exe. 

The 2nd option is to build using CMake, which is the default for non-Windows platforms, and the 'official' way to build JSBSim. In the CMake case a .lib file is first created and this is then later on linked in to also generate a .exe file.

The Visual Studio IDE from version 2017 onwards has included support in the IDE for building via CMake, see the following documentation which describes both Visual Studio options for building, i.e. via CMake and via the solution/project file.

https://jsbsim-team.github.io/jsbsim-reference-manual/mypages/quickstart-building-using-visualstudio/

So in your case you'll need to make use of the CMake option. You'll see in the CMake build documentation references to debug and release.

**Reply by bcoconni** [reply]
Another option has been suggested in issue #337 where we are considering distributing the compiled library with the headers files in the `setup.exe` package. That would avoid compiling the library yourself.

**Reply by sthelia** [reply]
Thanks! 
I did read that part of the reference manual, however with Visual Studio 2019 it looks a little bit different, so I couldn't follow the instructions exactly. I did manage to use the "open with CMake" option, but from there I had difficulty to change it to release. 

I might have missed something, I will have access to the windows computer again next week and will take another look to see if I can figure it out. If not, I'll post a screenshot to try to clarify my problem!

**Comment by bcoconni** [comment]
This issue has been moved to "Discussions". The github issues are reserved for reporting bugs or new feature requests.
This change is only motivated by project organization and will not affect the support provided. We'll gladly continue to answer your questions :smile: .

**Comment by sthelia** [comment]
Alright, I think I figured it out. It looks a bit different in VS2019, but I opened CMakeLists.txt file using File->Open->Cmake. Then I went in to CMakeSettings.json to change the configuration name from `x64-debug (default)` (maybe not needed, but made it easier to write the path later), and I changed to only build JSBSim and made sure it was release. 
I think the biggest difference was changing from building the entire JSBSim.vcxproj to just `JSBSim.exe (scr\JSBSim.exe)`, but I'm not completely sure. It looks like this when done: 
![jsbsim_vs_done](https://user-images.githubusercontent.com/52634491/112833830-8f165900-9097-11eb-9bd8-21d514cf48b0.png)

I'm still running in to problems compiling with mex, somehow the referencing of some variables doesn't seem to work. Not sure if this is a JSBSim problem or Matlab problem yet. (: 
![matlab_errors](https://user-images.githubusercontent.com/52634491/112834172-f502e080-9097-11eb-8118-107a35c2e706.png)

**Reply by bcoconni** [reply]
JSBSim needs to be linked to the libraries `wsock32.lib` and `ws2_32.lib`. Hopefully this should solve all the errors occurring with the class `JSBSim::FGfdmSocket`

https://github.com/JSBSim-Team/jsbsim/blob/375f5be0fdfeb210f4cf66f938e3e6fd426c2966/src/CMakeLists.txt#L28-L29

**Comment by seanmcleod** [comment]
The linker errors you listed above are because the JSBSim code includes calls to network/socket APIs but when you're building and linking your S-Function code via mex you're not including linker references to ws2_32.lib.

**Comment by sthelia** [comment]
Two fast replies, thanks a lot! It solved that problem right away. 

I ended up using the library file that I got from installing instead of building, because then I was provided with a folder with library file and an include folder with header files. 

Now I have it working both on Windows and Linux. :D

**Reply by agodemar** [reply]
@sthelia I'm glad you solved your issues. I would ask you, please, to share your settings, configuration, and howto-s. This would enable us to enhance our documentation and support for many users who would like to use right away the latest JSBSim with the latest versions of Matlab & Simulink. Thanks

**Reply by sthelia** [reply]
@agodemar I'll be happy to share my solutions once they are done and cleaned up! (: It's still a work in progress, so I'm expecting it some time in June.

**Reply by agodemar** [reply]
@sthelia That would be great, thanks!

**Comment by sthelia** [comment]
Hi again,
I am now wrapping up my thesis work. I would be happy to share my work, but to do so I would like to get some assistance on how to best prepare it for upload to github (as I have never used git before), and how the files should be organized. I'm also thinking it would need some kind of user guide, so some feedback/help with that would also be appreciated. 

If someone with Matlab and Simulink (I run 2020b, but earlier versions could work) would be willing to help me out with this, I'll contact you. I've used it on both Linux and Windows, so either should work. 

Thanks!

**Reply by agodemar** [reply]
Hi @sthelia 
That's great to hear about your work. I'll be happy to assist you and merge your code to the official JSBSim repo.

As I understand, you have been working on your local version of JSBSim, i.e. a copy of the official repository on your hard disk. You sorted out all the issues and made your tests on it, is it correct?
If so, and you want to offer now your code to the community, some additional steps are necessary. There could be different ways to do this. The following is the way I would go for:

- First of all, you should make your own fork of the official JSBSim repository, so that you have in your Github space a clone of JSBSim. [To learn more on what it means to make a fork, see here](https://docs.github.com/en/github/getting-started-with-github/quickstart/fork-a-repo).
- Then, you should install Git on your machine. [On Windows you can download it from here](https://git-scm.com/download/win).
- Once, you have Git working on your system, you will clone your forked repo from GitHub to your local hard disk.
- Then, you will select the source files you have modified in your local working version of JSBSim and copy them to the local cloned repository.
- At this point you should make sure that the tests in the cloned/amended local repository run as expected.
- Now the cloned/amended local repository is ready to be _pushed_ to your GitHub fork.
- Finally, you can make a pull request from your GitHub fork to the master branch of JSBSim GitHub repository. [To learn more on pull requests go here](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request).

@bcoconni do you have comments on this? I guess we can create a branch in JSBSim official repo and make a pull request there.

@sthelia As an alternative, you can share privately a folder with your work and I can manage the whole process.

**Reply by bcoconni** [reply]
> @bcoconni do you have comments on this? I guess we can create a branch in JSBSim official repo and make a pull request there.

As you explained, the best way is to follow GitHub philosophy and submit the code as a pull request via a fork. That way we can take advantage of GitHub review code features as well as GitHub actions which automates code compilation and tests run.

> - Then, you should install Git on your machine. [On Windows you can download it from here.](https://git-scm.com/download/win)

Yes and alternatively, if @sthelia feels more comfortable with GUI tools there is [GitHub desktop](https://desktop.github.com/) which provides an interface to git and saves you the trouble of managing obscure command line commands.

**Reply by sthelia** [reply]
I do think the desktop option would be the way to go for me! I have only used git to download, but not to modify anything, a GUI is probably easier to understand. I will take a look at it. 

So once I have the tests working, and I push to my branch, will that be visible for everyone else? Figure it would be good to get some comments on it before it getts pulled to the master.

**Reply by seanmcleod** [reply]
@sthelia take a look at:

https://careerkarma.com/blog/git-pull-request/

https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork

**Reply by bcoconni** [reply]
> So once I have the tests working, and I push to my branch, will that be visible for everyone else? 

Yes if your repository is public (and you provide us with a link :wink:)
> Figure it would be good to get some comments on it before it getts pulled to the master.

Actually that's the whole point of a pull request (PR) to get comments and make updates, so feel free to submit a PR early. If you push commits later to the branch you pulled they will automatically update the PR and we can iterate together on the final version of your patch before committing it to `master`.

**Comment by seanmcleod** [comment]
Just to confirm, my understanding is you've implemented some code to wrap JSBSim up as a Matlab S-Function?

https://www.mathworks.com/help/simulink/sfg/example-of-a-basic-c-mex-s-function.html

**Reply by sthelia** [reply]
Correct. I've basically updated and modified the S-function and interface from [here](https://github.com/podhrmic/JSBSim-Matlab), which I think is based on the S-function that @agodemar first developed. 
However, it is quite a large modification from that repo since I have completely skipped the GUI and just worked on the S-function block.

**Comment by seanmcleod** [comment]
In terms of organizing the code I would suggest just as we have python and julia wrappers in their own root level directories that we add a matlab directory.

Now the s-function wrapper builds a shared lib/dll but requires access to matlab's header files, libs and tools like `mex`. So for 99% of JSBSim developers building JSBSim their builds will skip building the matlab directory. So we would have something similar to what is done for julia, i.e. detect whether matlab is installed in order to decide whether to build the matlab directory. 

https://github.com/JSBSim-Team/jsbsim/blob/6e167e49606d2355139f31b34cbde6d9cfa4340c/julia/CMakeLists.txt#L1-L13

Ideally at a minimum I guess some documentation even just as a readme.md in the matlab directory explaining the functionality.

**Reply by bcoconni** [reply]
That would be great !

**Reply by sthelia** [reply]
That sounds good. I already have most of it in a root level directory so that shouldn't need much modification from my side. Building the S-function requires a mex-command, yes. I have done this manually in Matlab, it might be possible to do from CMake but I don't know how. I'll try to organise everything so you can see it and comment if something needs changing. 

I have also introduced some small modifications to include another operating system, QNX Neutrino. Would that be relevant to keep? 

Will definitely include some instructions on how to run it.

**Reply by bcoconni** [reply]
> I have also introduced some small modifications to include another operating system, QNX Neutrino. Would that be relevant to keep?

The normal process for PRs is to introduce one feature at a time. Ideally you should submit the mods for QNX Neutrino in a separate PR, is it feasible ?
If it's too much trouble for you, just keep your code as it is and we will see how this can be managed.

**Comment by seanmcleod** [comment]
Looks like there is a matlab option for `find_package`.

https://cmake.org/cmake/help/latest/module/FindMatlab.html

**Comment by agodemar** [comment]
There is also the possibility to integrate **GitHub Actions**, our automatic build workflow, with Matlab: [https://it.mathworks.com/help/matlab/matlab_prog/continuous-integration-with-matlab-on-ci-platforms.html](https://it.mathworks.com/help/matlab/matlab_prog/continuous-integration-with-matlab-on-ci-platforms.html).

Learn more here: [https://github.com/matlab-actions/overview/blob/main/README.md](https://github.com/matlab-actions/overview/blob/main/README.md)

**Reply by bcoconni** [reply]
Nice !

I'd rather use TravisCI since we already have accounts and experience on this CI platform (file [.travis.yml](https://github.com/JSBSim-Team/jsbsim/blob/bf97f722db022e8236df975bb3073d302874ec41/.travis.yml)). I'm actually surprised that they provide free access to Matlab: [Currently, MATLAB builds are available only for public projects in Linux® environments.](https://docs.travis-ci.com/user/languages/matlab/)

**Reply by bcoconni** [reply]
Ha, wait a minute: it is also available on GitHub actions ! How come that MathWorks provides this free access accross CI platforms to their product ?

**Reply by agodemar** [reply]
> 
> 
> Ha, wait a minute: it is also available on GitHub actions ! How come that MathWorks provides this free access accross CI platforms to their product ?

I guess it's a move to support Matlab dissemination in many engineering fields, where this technology is a standard _de facto_.

**Reply by bcoconni** [reply]
There are some caveats however: Matlab is only available on Linux (both on Travis CI and on GitHub) and MATLAB Coder™ and MATLAB Compiler™ are not installed. Is that a showstopper ?

**Reply by agodemar** [reply]
> 
> 
> There are some caveats however: Matlab is only available on Linux (both on Travis CI and on GitHub) and MATLAB Coder™ and MATLAB Compiler™ are not installed. Is that a showstopper ?

No, it's not. I think we will need the [tool that creates a Mex file](https://it.mathworks.com/help/matlab/ref/mex.html). To learn more see here: [https://it.mathworks.com/help/matlab/matlab_external/build-c-mex-programs.html](https://it.mathworks.com/help/matlab/matlab_external/build-c-mex-programs.html)

[MATLAB® Compiler™ enables you to share MATLAB programs as standalone applications and web apps](https://it.mathworks.com/help/compiler/index.html).

[MATLAB® Coder™ generates C and C++ code from MATLAB code for a variety of hardware platforms, from desktop systems to embedded hardware](https://it.mathworks.com/help/coder/).

**Reply by bcoconni** [reply]
Good ! I have not used Matlab since the 20th century (it was on a Sun/Solaris workstation, oh my !) so I am lagging a bit behind on that topic :smile:

**Reply by agodemar** [reply]
The current Matlab ecosystem is a great environment to work with. Sure, it's a commercial product, but in terms of features and versatility it's a mature software. That's why it is used in so many fields. 
I am glad there's the chance to test GitHub's continuous integration feature with JSBSim. Also because JSBSim can be integrated more easily in larger projects based on Matlab/Simulink.

**Comment by seanmcleod** [comment]
Yep, according to this link https://www.mathworks.com/solutions/continuous-integration.html 

> Use CI to automatically test and verify MATLAB® code and Simulink® models, **build MEX files**, generate code, and package toolboxes or libraries for deployment.

The CI tools do support the generation of MEX files. Although as you mentioned they're only available on Linux. So anyone using Matlab on Windows would need to build the s-function MEX file themselves.

**Reply by agodemar** [reply]
In terms of testing it's fair enough. A mex file can be generated and a Matlab script (an M-file) can then be run to test it.

**Reply by bcoconni** [reply]
Done ! I have just updated @sthelia's PR to build the S-Function and run a test `TestJSBSim.m` (see [there](https://github.com/JSBSim-Team/jsbsim/runs/2864424988?check_suite_focus=true))
